### PR TITLE
Practice Exercise on Testing

### PR DESCRIPTION
**Pull request by CheezItMan**
## Description
I'm trying to build a few practice exercises involving:

1.  Writing code using loops and arrays, specifically processing a collection of objects
1.  Reading Tests and writing code to match
1.  Writing Tests 

I'll be adding to this today & tomorrow, but this is the first thing to look at. It covers writing tests and reading the tests of another student, and a bit of work with arrays. 

I'd like to see if I we could pair some of our struggling students to do these exercises because I think they might benefit from working together on something small. 

## Todos
Are any of these items still in progress?
- [ ] Exercise specifically using arrays of objects and loops to process them
- [ ] Exercise on reading tests and writing code to match (also using loops and arrays).

## Feedback Requested
Ideas are **VERY** welcome

### CODE REVIEW

File: 02-intermediate-ruby/exercises/more-testing-practice.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+In this exercise you will look at a problem description and write a set of tests to evaluate a solution.  Then you will review the tests of another classmate and make suggestions for improvement.  Lastly you will write solution code to satisfy your tests.
+
+## Exercise Goal
+
+The goal of this set of exercises is to:
+
+*   Practice writing tests
+*   Review the tests written by another and look for:
+  -Testing all edge-cases
```
Comment (kariabancroft) [inline 02-intermediate-ruby/exercises/more-testing-practice.md]: These two bullets are not formatted correctly

File: 02-intermediate-ruby/exercises/more-testing-practice.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+## Flatten Array Problem Description
+
+This method should take a nested list and return a single flattened list with all values except nil/null.
+
+The challenge is to write a function that accepts an arbitrarily-deep nested list-like structure and returns a flattened structure without any nil/null values.
+
+For Example:
+
+Ex 1:
+input: `[1,[2,3,null,4],[null],5]`
```
Comment (kariabancroft) [inline 02-intermediate-ruby/exercises/more-testing-practice.md]: I think we should use `nil` if we're planning on giving this to them within the context of their time in Ruby. Then we could avoid any confusion with the nil/null section above.

File: 02-intermediate-ruby/exercises/more-testing-practice.md [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+## Flatten Array Problem Description
+
+This method should take a nested list and return a single flattened list with all values except nil/null.
+
+The challenge is to write a function that accepts an arbitrarily-deep nested list-like structure and returns a flattened structure without any nil/null values.
+
+For Example:
+
+Ex 1:
```
Comment (kariabancroft) [inline 02-intermediate-ruby/exercises/more-testing-practice.md]: The space between the example header and input is flattened, so if you add another carriage return you should be able to see it

File: 02-intermediate-ruby/exercises/more-testing-practice.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+output: `[1,2,3,4,5]`
+
+Ex 2:
+input: `['Peter', ['Paul', [ 'Mary' ]]]]`
+
+output: `['Peter', 'Paul', 'Mary']`
+
+## Stage 1:  Write the Tests
+
+Write a set of tests to evaluate a method named `flatten_array` which takes a single array as a parameter and returns that array flattened like above.  It should also remove any null values.
```
Comment (kariabancroft) [inline 02-intermediate-ruby/exercises/more-testing-practice.md]: Do you think it would be worthwhile to add a "Think about" section where we provide a few basic prompts to help them think about the types of things they should be testing (without giving it away)?

File: 02-intermediate-ruby/exercises/more-testing-practice.md [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+
+## Stage 3:  Test Review
```
Comment (kariabancroft) [inline 02-intermediate-ruby/exercises/more-testing-practice.md]: Should this be Stage 2 instead or as we missing a set of steps you want to include?

File: 02-intermediate-ruby/exercises/more-testing-practice.md [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+With a classmate review your written tests and look for:
+
+*   Edge-cases uncovered
+*   Return values tested correctly
+*   Opportunities to DRY-up code
+
+
+## Stage 4:  Refactor and Solution
+
+Finally refactor your tests from your partner's suggestions and write a `flatten_array` method to satisfy your tests.  You may not use the Array class' built-in method to flatten the array.
```
Comment (kariabancroft) [inline 02-intermediate-ruby/exercises/more-testing-practice.md]: I'd love to see as part of this step a reflection on whether or not the tests that they came up with ended up covering the code that they decided to write. I think that is also an interesting this to examine.

**Comment by CheezItMan** [comment]
@kariabancroft Thanks, I've revised it into more of an exercise in identifying where code will fail the tests, rather than solving the problem.  I'll add a solution later for those who want to see one.
